ldms/src/sampler/zfs_leafvdevs/zfs_leafvdevs.c

```diff
@@ -0,0 +1,525 @@
+				pool_name);
+	        /* zpool state  1 */
+		ldms_record_array_set_str(record_instance, vdev_metric_ids[1],
+				zpool_state_to_name((vdev_state_t)vs->vs_state, (vdev_aux_t)vs->vs_aux));
+	        /* vdevname     2 */
+		ldms_record_array_set_str(record_instance, vdev_metric_ids[2],
+						vdev_name);
+	        /* alloc        3 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[3], vs->vs_alloc);
+	        /* free         4 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[4], vs->vs_space - vs->vs_alloc);
+		/* size         5 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[5], vs->vs_space);
+		/* read_bytes   6 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[6], vs->vs_bytes[ZIO_TYPE_READ]);
+		/* iread_errors 7 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[7], vs->vs_read_errors);
+		/* read_ops     8 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[8], vs->vs_ops[ZIO_TYPE_READ]);
+		/* write_bytes  9 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[9], vs->vs_bytes[ZIO_TYPE_WRITE]);
+		/* write_errors 10 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[10], vs->vs_write_errors);
+		/* write_ops    11 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[11], vs->vs_ops[ZIO_TYPE_WRITE]);
+		/* checksum errors 12 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[12], vs->vs_checksum_errors);
+		/* fragmentation 13 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[13], vs->vs_fragmentation);
+		/* initialization errors 14 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[14], vs->vs_initialize_errors);
+	}
+
+        return (rc);
+}
+
+
+/*
+ * recursive stats printer
+ */
+static int get_recursive_stats(stat_printer_f func, nvlist_t *nvroot,
+    const char *pool_name, const char *parent_name, int descend)
+{
+        uint_t c, children;
+        nvlist_t **child;
+        char vdev_name[256];
+        int err;
+
+        /*err = func(nvroot, pool_name, parent_name);
+        if (err)
+                return (err);
+	*/
+        if (descend && nvlist_lookup_nvlist_array(nvroot, ZPOOL_CONFIG_CHILDREN,
+            &child, &children) == 0) {
+                (void) strlcpy(vdev_name, get_vdev_name(nvroot, parent_name),
+                    sizeof (vdev_name));
+
+                for (c = 0; c < children; c++) {
+			get_recursive_stats(func, child[c], pool_name,
+						vdev_name, descend);
+		}
+        } else {
+		err = func(nvroot, pool_name, parent_name);
+		if (err)
+			return (err);
+	}
+        return (0);
+}
+
+
+/*
+ * call-back to print the stats from the pool config
+ *
+ * Note: if the pool is broken, this can hang indefinitely and perhaps in an
+ * unkillable state.
+ */
+
+int get_leaf_stats(zpool_handle_t *zhp, void *data)
+{
+        uint_t          c;
+        int             err;
+        boolean_t       missing;
+        nvlist_t       *config, *nvroot;
+        vdev_stat_t    *vs;
+        struct timespec tv;
```

Comment:
Unused

Reply:
left overs from cut and paste. No excuse.